tests/Integration/PrestaShopBundle/Controller/Sell/Customer/Address/AddressControllerTest.php

```
@@ -0,0 +1,168 @@
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace Tests\Integration\PrestaShopBundle\Controller\Sell\Customer\Address;
+
+use Country;
+use PrestaShop\PrestaShop\Core\Exception\TypeException;
+use Tests\Integration\PrestaShopBundle\Controller\GridControllerTestCase;
+use Tests\Integration\PrestaShopBundle\Controller\TestEntityDTO;
+
+class AddressControllerTest extends GridControllerTestCase
+{
+    private $countryId;
+
+    public function __construct($name = null, array $data = [], $dataName = '')
+    {
+        parent::__construct($name, $data, $dataName);
+
+        $this->gridRoute = 'admin_addresses_index';
+        $this->testEntityName = 'address';
+        $this->deleteEntityRoute = 'admin_addresses_delete';
+    }
+
+    /**
+     * Tests all provided entity filters
+     * All filters are tested in one test make tests run faster
+     *
+     * @throws TypeException
+     */
+    public function testAddressFilters(): void
+    {
+        foreach ($this->getTestFilters() as $testFilter) {
+            $this->assertFiltersFindOnlyTestEntity($testFilter);
+        }
+    }
+
+    /**
+     * @return array
+     */
+    protected function getTestFilters(): array
+    {
+        return [
+            ['address[id_address]' => $this->getTestEntity()->getId()],
+            ['address[firstname]' => 'firstname'],
+            ['address[lastname]' => 'testLa'],
+            ['address[address1]' => 'address1'],
+            ['address[postcode]' => '11111'],
+            ['address[city]' => 'stcity'],
+            ['address[id_country]' => $this->countryId],
+        ];
+    }
+
+    /**
+     * @return TestEntityDTO
+     */
+    protected function getTestEntity(): TestEntityDTO
+    {
+        return new TestAddressDTO(
```

Here as well you created a `TestAddressDTO` object this means we will have to create a DTO for each tested controller, we already have a mountain of DTO in the software that are nearly clones (command, query result, domain logic DTO all are the same except not in the same namespace..) so lets' avoid adding some new ones in the test namespace

Instead you can make a `TestEntityDTO` class which is dynamic and accepts several parameters as input, if we want to still have some bit of checking we would make the input as an array `['firstName' => 'testfirstanme', 'lastName' => 'testlastname', ...]` and then we can use `__get()` method to have magic getters so that `getFIrstName()` will return the field from the input array. This way if a getter is called that doesn't match the input array we can throw an exception

---

Good point, haven't used magic methods before but it seemed pretty straightward. Implemented the changes.